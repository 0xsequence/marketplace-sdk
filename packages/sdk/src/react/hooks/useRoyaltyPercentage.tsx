import { type ChainId, type QueryArg, collectableKeys } from '@internal';
import { queryOptions, useQuery } from '@tanstack/react-query';
import type { Hex } from 'viem';
import { getContract } from 'viem';
import { EIP2981_ABI } from '../../utils/abi/abi/standard/EIP2981';
import { getPublicRpcClient } from '../../utils/get-public-rpc-client';

type UseRoyaletyPercentageArgs = {
	chainId: ChainId;
	collectionAddress: string;
	collectibleId: string;
} & QueryArg;

const fetchRoyaletyPercentage = async (args: UseRoyaletyPercentageArgs) => {
	const publicClient = getPublicRpcClient(args.chainId);

	const contract = getContract({
		address: args.collectionAddress as Hex,
		abi: EIP2981_ABI,
		client: publicClient,
	});

	try {
		const [_, royaltyPercentage] = await contract.read.royaltyInfo([
			BigInt(args.collectibleId),
			100n,
		]);

		return royaltyPercentage;
	} catch {
		//TODO: dont swallow errors
		return 0n;
	}
};

export const royaletyPercentageOptions = (args: UseRoyaletyPercentageArgs) =>
	queryOptions({
		...args.query,
		queryKey: [...collectableKeys.royaltyPercentage, args],
		queryFn: () => fetchRoyaletyPercentage(args),
	});

export const useRoyaltyPercentage = (args: UseRoyaletyPercentageArgs) => {
	return useQuery(royaletyPercentageOptions(args));
};
