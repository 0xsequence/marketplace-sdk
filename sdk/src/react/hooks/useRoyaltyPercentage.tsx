import { useReadContract } from 'wagmi';
import { z } from 'zod';
import { EIP2981_ABI } from '../../utils';
import { AddressSchema, ChainIdSchema, QueryArgSchema } from '../_internal';

const UseRoyaltyPercentageSchema = z.object({
	chainId: ChainIdSchema.pipe(z.coerce.string()),
	collectionAddress: AddressSchema,
	collectibleId: z.string(),
	query: QueryArgSchema.optional(),
});

type UseRoyaltyPercentageArgs = z.infer<typeof UseRoyaltyPercentageSchema>;

export const useRoyaltyPercentage = (args: UseRoyaltyPercentageArgs) => {
	const result = useReadContract({
		abi: EIP2981_ABI,
		address: args.collectionAddress,
		functionName: 'royaltyInfo',
		args: [BigInt(args.collectibleId), BigInt(100)],
		chainId: Number(args.chainId),
	});

	return result;
};
