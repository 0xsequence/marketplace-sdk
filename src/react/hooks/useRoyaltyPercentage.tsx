import {
	AddressSchema,
	ChainIdSchema,
	QueryArgSchema,
	collectableKeys,
} from '@internal';
import { queryOptions, useQuery } from '@tanstack/react-query';
import type { Hex } from 'viem';
import { getContract } from 'viem';
import { z } from 'zod';
import { EIP2981_ABI } from '../../utils/abi/abi/standard/EIP2981';
import { getPublicRpcClient } from '../../utils/get-public-rpc-client';

const UseRoyaletyPercentageSchema = z.object({
	chainId: ChainIdSchema.pipe(z.coerce.string()),
	collectionAddress: AddressSchema,
	collectibleId: z.string(),
	query: QueryArgSchema,
});

type UseRoyaletyPercentageArgs = z.infer<typeof UseRoyaletyPercentageSchema>;

const fetchRoyaletyPercentage = async (args: UseRoyaletyPercentageArgs) => {
	const parsedArgs = UseRoyaletyPercentageSchema.parse(args);
	const publicClient = getPublicRpcClient(parsedArgs.chainId);

	const contract = getContract({
		address: parsedArgs.collectionAddress as Hex,
		abi: EIP2981_ABI,
		client: publicClient,
	});

	try {
		const [_, royaltyPercentage] = await contract.read.royaltyInfo([
			BigInt(args.collectibleId),
			100n,
		]);

		return royaltyPercentage;
	} catch {
		//TODO: dont swallow errors
		return 0n;
	}
};

export const royaletyPercentageOptions = (args: UseRoyaletyPercentageArgs) =>
	queryOptions({
		...args.query,
		queryKey: [...collectableKeys.royaltyPercentage, args],
		queryFn: () => fetchRoyaletyPercentage(args),
	});

export const useRoyaltyPercentage = (args: UseRoyaletyPercentageArgs) => {
	return useQuery(royaletyPercentageOptions(args));
};
