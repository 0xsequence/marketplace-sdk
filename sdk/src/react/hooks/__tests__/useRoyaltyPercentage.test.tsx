import { renderHook, waitFor } from '@test';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { useRoyaltyPercentage } from '../useRoyaltyPercentage';

describe('useRoyaltyPercentage', () => {
	const mockAddress = '0x1234567890123456789012345678901234567890' as const;
	const mockArgs = {
		chainId: '1',
		collectionAddress: mockAddress,
		collectibleId: '1',
		query: {},
	};

	//const mockRoyaltyAmount = 500n; // 5% royalty

	beforeEach(() => {
		// Reset all mocks before each test
		vi.resetAllMocks();
	});

	it.skip('should fetch royalty percentage successfully', async () => {
		const { result } = renderHook(() => useRoyaltyPercentage(mockArgs));

		// Initially loading
		expect(result.current.isLoading).toBe(true);
		expect(result.current.data).toBeUndefined();

		// Wait for data to be loaded
		await waitFor(() => {
			expect(result.current.isLoading).toBe(false);
		});

		// Verify the data matches our mock
		expect(result.current.data).toBeDefined();
		// expect(result.current.data).toBe(mockRoyaltyAmount);
		expect(result.current.error).toBeNull();
	});

	it('should handle contract read error gracefully', async () => {
		const { result } = renderHook(() => useRoyaltyPercentage(mockArgs));

		await waitFor(() => {
			expect(result.current.isLoading).toBe(false);
		});

		// Should return 0 as specified in the hook implementation
		expect(result.current.data).toBe(0n);
		expect(result.current.isError).toBe(false);
	});

	it('should validate input parameters', async () => {
		// Using undefined as an invalid chain ID - this will fail Zod's string coercion
		const invalidArgs = {
			...mockArgs,
			chainId: undefined,
		};

		// @ts-expect-error - invalid args
		const { result } = renderHook(() => useRoyaltyPercentage(invalidArgs));

		// Wait for the query to complete
		await waitFor(() => {
			expect(result.current.isLoading).toBe(false);
		});

		// The query should fail with a validation error
		expect(result.current.isError).toBe(true);
		expect(result.current.error).toBeDefined();
	});

	it.skip('should cache the royalty data', async () => {
		const { result, rerender } = renderHook(() =>
			useRoyaltyPercentage(mockArgs),
		);

		await waitFor(() => {
			expect(result.current.isLoading).toBe(false);
		});

		// const mockClient = getPublicRpcClient('1') as MockPublicClient;
		// const mockReadContract = mockClient.readContract as Mock;
		// Record the number of calls to readContract
		// const initialCalls = mockReadContract.mock.calls.length;

		// Trigger a rerender
		rerender();

		// Should have data immediately from cache
		expect(result.current.isLoading).toBe(false);
		// expect(result.current.data).toBe(mockRoyaltyAmount);

		// Verify no additional contract calls were made
		// const finalCalls = mockReadContract.mock.calls.length;
		// expect(finalCalls).toBe(initialCalls);
	});
});
