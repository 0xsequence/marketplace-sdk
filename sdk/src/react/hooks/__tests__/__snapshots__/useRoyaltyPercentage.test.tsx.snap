// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useRoyaltyPercentage > should fetch royalty percentage successfully 1`] = `
[
  "0x8caA7E1431C5ad8583aE1734B61A41915Bf26f27",
  5n,
]
`;
